### PR TITLE
XWIKI-18599: Replace the Release live table with live data

### PR DESCRIPTION
**Pull request by manuelleduc**
https://jira.xwiki.org/browse/XWIKI-18599

### CODE REVIEW

File: xwiki-platform-core/xwiki-platform-release/xwiki-platform-release-ui/src/main/resources/Release/WebHome.xml [hunk 1]
```diff
@@ -63,19 +59,29 @@
...
+         'id': 'version', 'displayer': 'link'
+       },
+       {
+         'id': 'releaseManagers', 'displayer': 'xClassProperty'
+       },
+       {
+         'id': 'releaseDate', 'displayer': 'xClassProperty'
+       },
+       {
+         'id': 'releaseNotes', 'displayer': 'xClassProperty'
```
Comment (manuelleduc) [inline xwiki-platform-core/xwiki-platform-release/xwiki-platform-release-ui/src/main/resources/Release/WebHome.xml:75]: Of course this PR cannot be merge until XClassPropertyDisplayer is merged.

**Comment by manuelleduc** [comment]
Of course this is missing the docker test update.

File: xwiki-platform-core/xwiki-platform-livedata/xwiki-platform-livedata-livetable/pom.xml [hunk 2]
```diff
@@ -57,6 +57,14 @@
...
       <artifactId>javax.servlet-api</artifactId>
     </dependency>
+    <!-- Runtime dependencies -->
+    <dependency>
+      <groupId>org.xwiki.platform</groupId>
+      <artifactId>xwiki-platform-livetable-ui</artifactId>
+      <version>${project.version}</version>
+      <type>xar</type>
+      <scope>runtime</scope>
+    </dependency>
```
Comment (manuelleduc) [inline xwiki-platform-core/xwiki-platform-livedata/xwiki-platform-livedata-livetable/pom.xml:67]: @mflorea can you confirm it is relevant to add this dependency to make `livedata-livetable` work in a minimal XS distribution (eg, a docker test distrib).
Especially for https://github.com/xwiki/xwiki-platform/blob/e698f4fd5ff0eac8d89ecad761e285af6e6efc2a/xwiki-platform-core/xwiki-platform-livedata/xwiki-platform-livedata-livetable/src/main/java/org/xwiki/livedata/internal/livetable/LiveTableLiveDataEntryStore.java#L135

File: xwiki-platform-core/xwiki-platform-release/xwiki-platform-release-test/xwiki-platform-release-test-docker/src/test/it/org/xwiki/release/test/ui/ReleaseIT.java [hunk 3]
```diff
@@ -76,9 +85,9 @@ public void testRelease()
...
 
         // Assert Livetable:
-        // - verify that the Translation has been applied by checking the Translated livetable column name
-        // - verify that the Livetable contains our new Release entry
-        LiveTableElement lt = homePage.getReleaseLiveTable();
-        Assert.assertTrue(lt.hasRow("Version", RELEASE_VERSION));
+        // - verify that the Livetable contains our new Release entry.
+        LiveDataElement liveDataElement = homePage.getReleaseLiveData();
+        assertTrue(liveDataElement.hasRow("Version", RELEASE_VERSION),
+            String.format("%s expected to be found in column Version.", RELEASE_VERSION));
```
Comment (manuelleduc) [inline xwiki-platform-core/xwiki-platform-release/xwiki-platform-release-test/xwiki-platform-release-test-docker/src/test/it/org/xwiki/release/test/ui/ReleaseIT.java:91]: Straightforward migration of the existing test, more assertions are needed here.

File: xwiki-platform-core/xwiki-platform-livedata/pom.xml [hunk 4]
```diff
@@ -38,5 +38,6 @@
     <module>xwiki-platform-livedata-rest</module>
     <module>xwiki-platform-livedata-war</module>
     <module>xwiki-platform-livedata-webjar</module>
+    <module>xwiki-platform-livedata-test</module>
```
Comment (manuelleduc) [inline xwiki-platform-core/xwiki-platform-livedata/pom.xml:41]: needs to be moved in the `integration-tests`profile

File: xwiki-platform-core/xwiki-platform-livedata/xwiki-platform-livedata-test/xwiki-platform-livedata-test-pageobjects/src/main/java/org/xwiki/livedata/test/po/LiveDataElement.java [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+    /**
+     * Gets the values currently displayed in the given column.
+     *
+     * @param columnName the column name
+     * @return the list of values found in the column
+     */
+    private List<WebElement> getValues(String columnName)
+    {
+        int columnIndex = getColumnIndex(columnName);
+        System.out.println("COLUMN INDEX " + columnIndex);
```
Comment (manuelleduc) [inline xwiki-platform-core/xwiki-platform-livedata/xwiki-platform-livedata-test/xwiki-platform-livedata-test-pageobjects/src/main/java/org/xwiki/livedata/test/po/LiveDataElement.java:100]: Shouldn't have been committed, will be removed.

File: xwiki-platform-core/xwiki-platform-release/xwiki-platform-release-test/xwiki-platform-release-test-docker/pom.xml [hunk 6]
```diff
@@ -67,47 +59,38 @@
...
-    <!-- Make sure we have the Logback SLF4J binding for client side logging for our tests and make sure other logging
-         systems go through SLF4J -->
     <dependency>
-      <groupId>ch.qos.logback</groupId>
-      <artifactId>logback-classic</artifactId>
-      <scope>test</scope>
+      <groupId>org.xwiki.platform</groupId>
+      <artifactId>xwiki-platform-localization-rest-default</artifactId>
+      <version>${project.version}</version>
+      <scope>runtime</scope>
```
Comment (mflorea) [inline xwiki-platform-core/xwiki-platform-release/xwiki-platform-release-test/xwiki-platform-release-test-docker/pom.xml:76]: Why is this needed here? Doesn't seem right.

File: xwiki-platform-core/xwiki-platform-release/xwiki-platform-release-test/xwiki-platform-release-test-docker/pom.xml [hunk 7]
```diff
@@ -67,47 +59,38 @@
...
     </dependency>
     <dependency>
-      <groupId>org.slf4j</groupId>
-      <artifactId>jcl-over-slf4j</artifactId>
-      <scope>test</scope>
+      <groupId>org.xwiki.platform</groupId>
+      <artifactId>xwiki-platform-livedata-rest</artifactId>
+      <version>${project.version}</version>
+      <scope>runtime</scope>
```
Comment (mflorea) [inline xwiki-platform-core/xwiki-platform-release/xwiki-platform-release-test/xwiki-platform-release-test-docker/pom.xml:82]: Same as above, this doesn't feel right. It shouldn't be listed here.

File: xwiki-platform-core/xwiki-platform-release/xwiki-platform-release-test/xwiki-platform-release-test-docker/pom.xml [hunk 8]
```diff
@@ -67,47 +59,38 @@
...
     </dependency>
     <dependency>
-      <groupId>org.slf4j</groupId>
-      <artifactId>log4j-over-slf4j</artifactId>
-      <scope>test</scope>
+      <groupId>org.xwiki.platform</groupId>
+      <artifactId>xwiki-platform-livedata-livetable</artifactId>
+      <version>${project.version}</version>
+      <scope>runtime</scope>
```
Comment (mflorea) [inline xwiki-platform-core/xwiki-platform-release/xwiki-platform-release-test/xwiki-platform-release-test-docker/pom.xml:88]: Same. This should probably be a dependency of the ``xwiki-platform-release-ui`` (if it uses the live data macro with the live table source).

File: xwiki-platform-core/xwiki-platform-release/xwiki-platform-release-test/xwiki-platform-release-test-docker/src/test/it/org/xwiki/release/test/ui/AllITs.java [hunk 9]
```diff
@@ -19,16 +19,22 @@
...
- * @since 13.0
+ * @since 13.4RC1
  */
-@RunWith(PageObjectSuite.class)
-public class AllITs
+@UITest
+class AllITs
 {
+    @Nested
+    @DisplayName("Overall AWM LiveTable test")
```
Comment (mflorea) [inline xwiki-platform-core/xwiki-platform-release/xwiki-platform-release-test/xwiki-platform-release-test-docker/src/test/it/org/xwiki/release/test/ui/AllITs.java:36]: "AWM"? What's the relation with AppWithinMinutes?

File: xwiki-platform-core/xwiki-platform-release/xwiki-platform-release-ui/pom.xml [hunk 10]
```diff
@@ -66,9 +66,8 @@
     <!-- We need the livetable for the Release Home page that lists all Releases -->
     <dependency>
       <groupId>org.xwiki.platform</groupId>
-      <artifactId>xwiki-platform-livetable-ui</artifactId>
+      <artifactId>xwiki-platform-livedata-macro</artifactId>
```
Comment (mflorea) [inline xwiki-platform-core/xwiki-platform-release/xwiki-platform-release-ui/pom.xml:69]: You need to keep the ``livetable-ui`` (for the default results page) and add the dependency to the live table source. This module basically uses the live data **macro** with the live table **source** with the default live table **results page**. So it depends on all 3.

File: xwiki-platform-core/xwiki-platform-release/xwiki-platform-release-ui/src/main/resources/Release/WebHome.xml [hunk 11]
```diff
@@ -63,19 +59,36 @@
...
+       }
+    ]
+  }
 })
-#livetable("releases" $columns $columnsProperties $options)
+
+#set($sourceParameters = $escapetool.url({
+  'className': 'ReleaseCode.ReleaseClass',
+  'selectedColumn': 'releaseDate',
+  'defaultOrder': 'desc',
```
Comment (mflorea) [inline xwiki-platform-core/xwiki-platform-release/xwiki-platform-release-ui/src/main/resources/Release/WebHome.xml:84]: You need to use the ``sort`` live data **macro parameter**. See https://extensions.xwiki.org/xwiki/bin/view/Extension/Live%20Data%20Macro/#HParameters .

File: xwiki-platform-core/xwiki-platform-release/xwiki-platform-release-ui/src/main/resources/Release/WebHome.xml [hunk 12]
```diff
@@ -63,19 +59,36 @@
...
+       },
+       {
+         'id': 'releaseManagers', 'displayer': 'html'
+       },
+       {
+         'id': 'releaseDate', 'displayer': 'html'
+       },
+       {
+         'id': 'releaseNotes', 'displayer': 'html'
+       }
```
Comment (mflorea) [inline xwiki-platform-core/xwiki-platform-release/xwiki-platform-release-ui/src/main/resources/Release/WebHome.xml:76]: These should not be needed. You can rely on the default displayer for object properties (default live data configuration coming from the live table bridge).

**Review by mflorea — CHANGES_REQUESTED** [review]

File: xwiki-platform-core/xwiki-platform-release/xwiki-platform-release-test/xwiki-platform-release-test-docker/src/test/it/org/xwiki/release/test/ui/AllITs.java (hunk 9 above)
Comment (manuelleduc) [reply xwiki-platform-core/xwiki-platform-release/xwiki-platform-release-test/xwiki-platform-release-test-docker/src/test/it/org/xwiki/release/test/ui/AllITs.java:36]: None, bad copy past I guess... sorry, fixing

File: xwiki-platform-core/xwiki-platform-release/xwiki-platform-release-ui/src/main/resources/Release/WebHome.xml [hunk 13]
```diff
@@ -63,19 +59,20 @@
...
-   "selectedColumn" : "releaseDate",
-   "defaultOrder" : "desc",
-   "rowCount": 15,
-   "translationPrefix":"release.livetable."
+
+#set($liveDataConfig = {
+  'meta': {
+    'propertyDescriptors': [
+       { 'id': 'version', 'displayer': 'link' },
+       { 'id': 'releaseDate', 'name': $services.localization.render('release.livetable.releaseDate') }
```
Comment (manuelleduc) [inline xwiki-platform-core/xwiki-platform-release/xwiki-platform-release-ui/src/main/resources/Release/WebHome.xml:67]: name needed to keep the label as `Release Date` instead of `Release Date` (the field name).

File: xwiki-platform-core/xwiki-platform-livedata/xwiki-platform-livedata-test/xwiki-platform-livedata-test-pageobjects/src/main/java/org/xwiki/livedata/test/po/LiveDataElement.java [hunk 14]
```diff
@@ -0,0 +1,141 @@
...
+    }
+
+    /**
+     * Checks if the a column contains a value.
+     *
+     * @param columnName the column name
+     * @param value the value to be found in the column
+     * @return {@code true} if the value is found in the column, {@code false} otherwise
+     */
+    public boolean hasRow(String columnName, String value)
```
[the remark this answers is not in the document]
Comment (manuelleduc) [reply xwiki-platform-core/xwiki-platform-livedata/xwiki-platform-livedata-test/xwiki-platform-livedata-test-pageobjects/src/main/java/org/xwiki/livedata/test/po/LiveDataElement.java]: done

File: xwiki-platform-core/xwiki-platform-livedata/xwiki-platform-livedata-test/xwiki-platform-livedata-test-pageobjects/src/main/java/org/xwiki/livedata/test/po/LiveDataElement.java [hunk 15]
```diff
@@ -0,0 +1,141 @@
...
+            .stream()
+            .map(WebElement::getText)
+            .anyMatch(it -> it.equals(value));
+    }
+
+    /**
+     * Waits until the table has content displayed and loaded. Do not use this method if you expect the livedata to be
+     * displayed without content.
+     */
+    public void waitUntilHasContentReady()
```
[the remark this answers is not in the document]
Comment (manuelleduc) [reply xwiki-platform-core/xwiki-platform-livedata/xwiki-platform-livedata-test/xwiki-platform-livedata-test-pageobjects/src/main/java/org/xwiki/livedata/test/po/LiveDataElement.java]: done

File: xwiki-platform-core/xwiki-platform-release/xwiki-platform-release-test/xwiki-platform-release-test-docker/src/test/it/org/xwiki/release/test/ui/ReleaseIT.java [hunk 16]
```diff
@@ -76,9 +85,9 @@ public void testRelease()
...
 
         // Assert Livetable:
-        // - verify that the Translation has been applied by checking the Translated livetable column name
-        // - verify that the Livetable contains our new Release entry
-        LiveTableElement lt = homePage.getReleaseLiveTable();
-        Assert.assertTrue(lt.hasRow("Version", RELEASE_VERSION));
+        // - verify that the Livetable contains our new Release entry.
+        TableLayoutElement liveDataElement = homePage.getReleaseLiveData().getTableLayout();
+        assertTrue(liveDataElement.hasRow("Version", RELEASE_VERSION),
+            String.format("%s expected to be found in column Version.", RELEASE_VERSION));
```
Comment (vmassol) [inline xwiki-platform-core/xwiki-platform-release/xwiki-platform-release-test/xwiki-platform-release-test-docker/src/test/it/org/xwiki/release/test/ui/ReleaseIT.java:91]: idea: add an `assertRow()` API in `TableLayoutElement`.

File: xwiki-platform-core/xwiki-platform-release/xwiki-platform-release-ui/src/main/resources/Release/WebHome.xml (hunk 13 above)
Comment (vmassol) [reply xwiki-platform-core/xwiki-platform-release/xwiki-platform-release-ui/src/main/resources/Release/WebHome.xml:67]: shouldn't that info be added in the file instead of in a comment in the PR? :)

PS: I don't understand you comment (both value seem to the same).

Comment (mflorea) [reply xwiki-platform-core/xwiki-platform-release/xwiki-platform-release-ui/src/main/resources/Release/WebHome.xml:67]: @manuelleduc from your comment I deduce that the field pretty name takes precedence over the specified translation prefix, which is not right. It should be the other way around: if the translation prefix is specified and the corresponding key has a value then it should be used, otherwise it should fallback on the field (translated) pretty name. This needs to be fixed in ``livetable.DefaultLiveDataConfigurationResolver``. It's not critical and can be fixed later. So please remove the ``releaseDate`` descriptor. Would be good to report an issue on the live data for the property name fallback issue.

**Review by mflorea — CHANGES_REQUESTED** [review]

File: xwiki-platform-core/xwiki-platform-release/xwiki-platform-release-ui/src/main/resources/Release/WebHome.xml (hunk 13 above)
Comment (manuelleduc) [reply xwiki-platform-core/xwiki-platform-release/xwiki-platform-release-ui/src/main/resources/Release/WebHome.xml:67]: Issue created https://jira.xwiki.org/browse/XWIKI-18628 thanks.
